projects/plugins/jetpack/extensions/plugins/ai-assistant-plugin/components/breve/utils/textEditingHelpers.js

```diff
@@ -0,0 +1,106 @@
+import { requestAnimationFrame } from './requestAnimationFrame';
+
+export const saveCaretPosition = ( containerEl, caretPositionRef ) => {
+	const selection = containerEl.ownerDocument.defaultView.getSelection();
+
+	if ( selection.rangeCount > 0 ) {
+		const range = selection.getRangeAt( 0 );
+		const preCaretRange = range.cloneRange();
+		preCaretRange.selectNodeContents( range.startContainer );
+		preCaretRange.setEnd( range.startContainer, range.startOffset );
+		const caretOffset = preCaretRange.toString().length;
+
+		caretPositionRef.current = {
+			container: range.startContainer,
+			offset: caretOffset,
+		};
+	}
+};
+
+export const restoreCaretPosition = ( containerEl, caretPositionRef ) => {
+	const { container, offset } = caretPositionRef.current || {};
+
+	if ( container && typeof offset === 'number' ) {
+		const range = containerEl.ownerDocument.createRange();
+		const selection = containerEl.ownerDocument.defaultView.getSelection();
+		range.setStart( container, offset );
+		range.collapse( true );
+		selection.removeAllRanges();
+		selection.addRange( range );
+	}
+};
+
+export const getChildTextNodes = element => {
+	const textNodes = [];
+
+	for ( let i = 0; i < element.childNodes.length; i++ ) {
+		const childNode = element.childNodes[ i ];
+
+		if ( childNode.nodeType === Node.TEXT_NODE ) {
+			textNodes.push( childNode );
+		} else if ( childNode.nodeType === Node.ELEMENT_NODE ) {
+			const childNodeTextNodes = getChildTextNodes( childNode );
+			textNodes.push( ...childNodeTextNodes );
+		}
+	}
+
+	return textNodes;
+};
```

Comment:
@madebynoam This should fix the bug where clicking on a text that starts in bold throws an error in the console. A block can have many child nodes and not all of them are text nodes.